lib/controllers/login.js

```diff
@@ -4,152 +4,171 @@ var _ = require('lodash');
 var extend = require('deep-extend');
 var stormpath = require('stormpath');
 var url = require('url');
-var uuid = require('uuid');
 
 var forms = require('../forms');
 var helpers = require('../helpers');
 var oauth = require('../oauth');
-var loginWithOAuthProvider = require('../helpers/login-with-oauth-provider');
 
 /**
- * This controller logs in an existing user.  If there are any errors, an
- * error page is rendered.  If the process succeeds, the user will be logged in
- * and redirected.
+ * Authenticate with a username and password.
  *
  * @method
+ * @private
  *
- * @param {Object} req - The http request.
- * @param {Object} res - The http response.
+ * @param {string} username - Username to authenticate with.
+ * @param {string} password - Password to authenticate with.
+ * @param {function} callback - Function to call when authenticated.
  */
-module.exports = function (req, res) {
-  var accepts = req.accepts(['html', 'json']);
-  var application = req.app.get('stormpathApplication');
+function authenticate(application, credentials, callback) {
   var authenticator = stormpath.OAuthPasswordGrantRequestAuthenticator(application);
-  var config = req.app.get('stormpathConfig');
-  var logger = req.app.get('stormpathLogger');
-  var nextUri = url.parse(req.query.next || '').path;
-  var formActionUri = (config.web.login.uri + (nextUri ? ('?next=' + nextUri) : ''));
-  var view = req.app.get('stormpathConfig').web.login.view;
-  var oauthStateToken = null;
-  var providerData = req.body && req.body.providerData;
-
-  function renderForm(form, options) {
-    if (options === undefined) {
-      options = {};
-    }
-
-    var hasSocialProviders = _.some(config.socialProviders, function (socialProvider) {
-      return socialProvider.enabled;
-    });
 
-    extend(options, {
-      form: form,
-      formActionUri: formActionUri,
-      oauthStateToken: oauthStateToken,
-      hasSocialProviders: hasSocialProviders
-    });
+  credentials = credentials || {};
+  credentials.username = credentials.login || credentials.username || '';
+  credentials.password = credentials.password || '';
 
-    helpers.render(req, res, view, options);
+  if (!credentials.username || !credentials.password) {
+    return callback(new Error('Invalid username or password.'));
```

Comment:
Thanks.  So the `Username or password cannot be null or empty` error is actually coming from the REST API.  I do prefer this error because it's more descriptive (it tells you that you are missing a field), albeit confusing now that we're calling this field `login` at the framework layer.  Blarg.  I'm opening a discussion [here](https://github.com/stormpath/stormpath-framework-spec/issues/36).  Let's leave this as-is for now and revisit per discussion output.  This doesn't hold up this PR.